core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -518,6 +528,51 @@ public ByteBufAllocator alloc() {
         return channel != null ? channel.alloc() : PooledByteBufAllocator.DEFAULT;
     }
 
+    /**
+     * Adds a hook which is invoked whenever this {@link DefaultClientRequestContext} is pushed to the
+     * {@link RequestContextStorage}. The {@link SafeCloseable} returned by {@code contextHook} will be called
+     * whenever this {@link RequestContext} is popped from the {@link RequestContextStorage}.
+     * This method is useful when you need to propagate a custom context in this {@link RequestContext}'s scope.
+     *
+     * <p>Note that this operation is highly performance-sensitive operation, and thus
+     * it's not a good idea to run a time-consuming task.
+     */
+    @UnstableApi
+    public void hook(Supplier<? extends SafeCloseable> contextHook) {
```

Comment:
How about adding this method to `NonWrappingRequestContext` so that `DefaultClientRequestContext` and `DefaultClientRequestContext` can share the logic?

Reply:
Nice suggestion.